docs/contracts/transient-storage.rst

```diff
@@ -0,0 +1,61 @@
+.. index:: ! transient storage
+
+.. _transient-storage:
+
+*****************
+Transient Storage
+*****************
+
+Transient storage is another data location besides memory, storage, calldata
+(and return-data and code) which was introduced alongside its respective opcodes
+``TSTORE`` and ``TLOAD`` by `EIP-1153 <https://eips.ethereum.org/EIPS/eip-1153>`_.
+This new data location behaves as a key-value store similar to storage with the main
+difference being that data in transient storage is not permanent, but is scoped to
+the current transaction only, after which it will be reset to zero. The values on
+transient storage are never deserialized from storage or serialized to storage,
+thus the gas cost of operating in transient storage is much cheaper,
+since it doesn't require disk access.
+
+Currently, the compiler can parse ``transient`` as a data location, however it is not
+defined as a keyword of the language yet. This means that the use of ``transient``
+is backwards-compatible and does not break previous code that eventually used it as an identifier.
+Note however that such use of ``transient`` as a data location is only allowed for
+:ref:`value type <value-types>` state variable declarations. Further support for
+other types as well as code generation is intended with upcoming releases.
+
+An expected canonical use case for transient storage is cheaper reentrancy locks,
+which can be readily implemented with the opcodes as showcased next.
+
+.. code-block:: solidity
+
+    // SPDX-License-Identifier: GPL-3.0
+    pragma solidity ^0.8.27;
+
+    contract Generosity {
+        mapping(address => bool) sentGifts;
+        bool transient locked;
+
+        modifier nonReentrant {
+            require(!locked, "Reentrancy attempt");
+            locked = true;
+            _;
+            // Unlocks the guard, making the pattern composable.
+            // After the function exits, it can be called again, even in the same transaction.
+            locked = false;
+        }
+
+        function claimGift() nonReentrant public {
+            require(address(this).balance >= 1 ether);
+            require(!sentGifts[msg.sender]);
+            (bool success, ) = msg.sender.call{value: 1 ether}("");
+            require(success);
+
+            // In a reentrant function, doing this last would open up the vulnerability
+            sentGifts[msg.sender] = true;
+        }
+    }
+
+.. note::
+    Given the caveats mentioned in the specification of EIP-1153, utmost care is
+    recommended for more advanced use cases of transient storage to preserve
+    the composability of your smart contract.
```

Comment:
I think this is still not detailed enough. Listing those caveats is very important. The note at the very least should directly link to the [Security Considerations](https://eips.ethereum.org/EIPS/eip-1153#security-considerations) because otherwise very few people will go out of their way to look it up.

I'd go as far so to directly quote the relevant bits from that section here:
> Smart contract developers should understand the lifetime of transient storage variables before use. Because transient storage is automatically cleared at the end of the transaction, smart contract developers may be tempted to avoid clearing slots as part of a call in order to save gas. However, this could prevent further interactions with the contract in the same transaction (e.g. in the case of re-entrancy locks) or cause other bugs, so smart contract developers should be careful to only leave transient storage slots with nonzero values when those slots are intended to be used by future calls within the same transaction. Otherwise, these opcodes behave exactly the same as `SSTORE` and `SLOAD`, so all the usual security considerations apply especially in regard to reentrancy risk.
>
> Smart contract developers may also be tempted to use transient storage as an alternative to in-memory mappings. They should be aware that transient storage is not discarded when a call returns or reverts, as is memory, and should prefer memory for these use cases so as not to create unexpected behavior on reentrancy in the same transaction. The necessarily high cost of transient storage over memory should already discourage this usage pattern. Most usages of in-memory mappings can be better implemented with key-sorted lists of entries, and in-memory mappings are rarely required in smart contracts (i.e. the author knows of no known use cases in production).

It would also be good to have examples, but we already have some in the blog post so maybe it's enough to link to that post here.

Reply:
I had copied and adapted the text and an example from our blog post. I think they are equivalent.
I included a phrase in the end pointing to the Security Considerations section of EIP-1153.
What do you think?